### DISCUSSION TITLE
How can i have shruggies?

### DISCUSSION THREAD

**Discussion by semi710** [fdw/rofimoji#181] [Q&A] [answered]
`¯\_(ツ)_/¯`?

**Comment by fdw** [comment]
They haven't been added to `rofimoji` yet, because I don't use them and no-one else volunteered to add them.

What you'd need is just a file with a "character" (i.e. the "shruggy"), followed by a space ` ` and a description; one character per line.
Then you can add them to `rofimoji` with `--files`. And I'd be happy to include them here, if you open a PR :)

**Reply by semi710** [reply]
```
\(＾▽＾\) Happy Feeling joyous
\(￣▽￣\)ノ Cheerful Wearing a broad smile
\(⌒‿⌒\) Smiling Radiating happiness
\(◕‿◕\) Grinning Brimming with joy
\(◠‿◠\) Cute Utterly charming
\(´・ω・`\) Sad Deeply melancholic
\(ಥ_ಥ\) Crying Overflowing with tears
\(；一_一\) Disappointed Completely disheartened
\(¬_¬\) Sceptical Highly skeptical
\(´･_･`\) Annoyed Visibly irritated
\(╯°□°）╯︵ ┻━┻ Angry Flipping tables in rage
\(ʘᗩʘ'\) Surprised Completely astonished
\(*^▽^*\) Excited Bursting with excitement
\(¬‿¬\) Flirty Being flirtatious
\(⊙_☉\) Confused Completely bewildered
\(ง'̀-'́\)ง Determined Showing strong determination
\(¬‿¬ \) Playful Up to mischief
\( ͡° ͜ʖ ͡°\) Lenny Face Exuding slyness
\(╥_╥\) Upset Deeply distressed
\(◔_◔\) Skeptical Full of distrust
\(ᵔᴥᵔ\) Bear Face Exuding friendliness
\(ó ì_í)=óò=(ì_í ò\) Angry Bear Ready to fight fiercely
\(っ◔◡◔\)っ Hug Wrapped in an embrace
\(◕‿-\) Sleepy Drifting into sleepiness
\(°ロ°\) !? Shocked Completely astonished
\(^ω^\) Kawaii Overflowing with cuteness
\(¬‿¬\) Smug Brimming with smugness
\(◉_◉\) Suspicious Maintaining a wary gaze
\(¬▂¬\) Frustrated Frustrated and aggravated
\(´ー｀\) Bored Totally uninterested
\(≧◡≦\) Wholesome Heartwarming and endearing
\(=^・^=\) Cat Face With a feline grace
\(¬‿¬\) Sarcastic Showing a touch of sarcasm
\(๑╹◡╹)ﾉ" Waving Waving a warm hello
\(⌐■_■\) Cool Maintaining a cool demeanor
\(>_<\) Embarrassed Feeling sheepish
\(óvò\) Content Completely content
\(≧ω≦\) Blushing Bashfully blushing
\(ㆁᴗㆁ✿\) Cute Flower Radiating delight
\(´▽`ʃ♡ƪ\) Love Overflowing with affection
\(๑>ᴗ<๑\) Happy Blushing Happily blushing
\(◜‿◝\) Innocent Totally innocent
\(つ✧ω✧)つ Sending Love Sending affectionate vibes
\(◞‸◟\) Indifferent Completely indifferent
\(ﾉ･ｪ･\)ﾉ Excited Hands Up Overflowing with excitement
\(｡♥‿♥｡\) Sweet Overflowing with tenderness
\(°Д°\) Shocked Completely astonished
\(ᵔᴥᵔ)ʕ•ᴥ•ʔ Bear Hug Overflowing with affection
\(ﾟДﾟ\) Surprised Completely amazed
\(´･ᴗ･`\) Curious Deeply curious
```

I have this list of kaomoji, but there is a slight issue as some of these moji's contain spaces in them
it becomes hard for rofimoji to differentiate between description and the moji

e.g `\( ͡° ͜ʖ ͡°\) Lenny Face Exuding slyness`
I think the files should be in `json` format that will make it easier to differentiate between moji and description.

And can open future possibilities for users to have their custom emoji set's e.g `👊🏽👊🏽 Dual Punch`

```json
[{
	"emoji": "👊🏽👊🏽",
	"description": "Dual Punch"
},
{
	"emoji": "\( ͡° ͜ʖ ͡°\)",
	"description": "Lenny Face Exuding slyness"
}]
```

There are two ways imo we can approach this.
Either we convert all `csv` to `json` or make rofimoji compatible with having `csv` and `json` files.

**Reply by semi710** [reply]
While replying, I noticed `json` will require escape handling for `\ backward slash`. 
```json
[{
	"emoji": "👊🏽👊🏽",
	"description": "Dual Punch"
},
{
	"emoji": "\\( ͡° ͜ʖ ͡°\\)",
	"description": "Lenny Face Exuding slyness"
}]
```
`\ -> \\`

**Reply by semi710** [reply]
https://github.com/vsedov/kaomoji_scrapper
a resource i found

**Reply by fdw** [reply]
Thanks for the resources, I'll have a look 🙂

In regards to spaces, there was already #162. In short, the problem is not only reading the file correctly, but also parsing the return value from `rofi`/`wofi` etc. For the first, I had tried using "real" csv files, but the reading performance was so bad that I dropped the approach. I haven't tested json yet, but I'm not too optimistic there, either. Would it already help to package kaomoji that don't contain spaces?

**Reply by semi710** [reply]
Concern regarding the reading performance is reasonable, and ig it will be the same for `json` as it was for real `csv`.
Addition of kaomoji that don't contain spaces will be better than nothing.

**Reply by semi710** [reply]
I Liked the approach https://github.com/Rishabh672003/dotfiles/blob/main/.config/hypr/scripts/kaomoji.sh
as he uses some kind of Unicode character to keep the difference instead of space.
if rofimoji can accept separator string like `rofimoji -s " " -s "<unicodeChar>"` and then can seprate the content of the file based on this?

![image](https://github.com/fdw/rofimoji/assets/60490474/f1d63ef3-33ab-4d59-8038-68aecb1b35a8)
here in the image you can see the unicode character used by him.

https://github.com/Rishabh672003/dotfiles/blob/main/.config/hypr/scripts/kaomoji.txt <- This is the list of kaomoji with those Unicode character as separator.

**Reply by semi710** [reply]
Or we can define in the config `file` with the type of separator `<unicode>` which defaults to `space`

**Reply by fdw** [reply]
A different idea (that needs to be tested, of course): We could use a non-breaking space in the kaomoji; then the normal space can work as a separator as before. And a non-breaking space makes a lot of sense in the kaomoji, too, as you don't want them to broken apart. What do you think?

**Reply by semi710** [reply]
@fdw here I have created a collection of kaomoji's where the `moji's` are having a non-breaking space ` `.
And descriptions are having normal spaces with separator as normal space ` `.
https://gist.github.com/niksingh710/5396cf634e807bf487a888abe8740ea1

For most of the part it is working as expected with no issue.
![image](https://github.com/fdw/rofimoji/assets/60490474/38e76f7b-6ac8-445d-8678-06698abf7d66)
just there are some entries `DDD` are listed, don't know why are they coming up

**Reply by fdw** [reply]
Awesome, thank you! 🙂

I'll try to have a look in the next days

**Comment by fdw** [comment, marked as answer]
I've opened ticket #182 for this
